src/torcontrol.cpp

```diff
@@ -0,0 +1,685 @@
+ * 'AUTH' and arguments 'METHODS=...'.
+ */
+static std::pair<std::string,std::string> SplitTorReplyLine(const std::string &s)
+{
+    size_t ptr=0;
+    std::string type;
+    while (ptr < s.size() && s[ptr] != ' ') {
+        type.push_back(s[ptr]);
+        ++ptr;
+    }
+    if (ptr < s.size())
+        ++ptr; // skip ' '
+    return make_pair(type, s.substr(ptr));
+}
+
+/** Parse reply arguments in the form 'METHODS=COOKIE,SAFECOOKIE COOKIEFILE=".../control_auth_cookie"'.
+ */
+static std::map<std::string,std::string> ParseTorReplyMapping(const std::string &s)
+{
+    std::map<std::string,std::string> mapping;
+    size_t ptr=0;
+    while (ptr < s.size()) {
+        std::string key, value;
+        while (ptr < s.size() && s[ptr] != '=') {
+            key.push_back(s[ptr]);
+            ++ptr;
+        }
+        if (ptr == s.size()) // unexpected end of line
+            return std::map<std::string,std::string>();
+        ++ptr; // skip '='
+        if (ptr < s.size() && s[ptr] == '"') { // Quoted string
+            ++ptr; // skip '='
+            bool escape_next = false;
+            while (ptr < s.size() && (!escape_next && s[ptr] != '"')) {
+                escape_next = (s[ptr] == '\\');
+                value.push_back(s[ptr]);
+                ++ptr;
+            }
+            if (ptr == s.size()) // unexpected end of line
+                return std::map<std::string,std::string>();
+            ++ptr; // skip closing '"'
+            /* TODO: unescape value - according to the spec this depends on the
+             * context, some strings use C-LogPrintf style escape codes, some
+             * don't. So may be better handled at the call site.
+             */
+        } else { // Unquoted value. Note that values can contain '=' at will, just no spaces
+            while (ptr < s.size() && s[ptr] != ' ') {
+                value.push_back(s[ptr]);
+                ++ptr;
+            }
+        }
+        if (ptr < s.size() && s[ptr] == ' ')
+            ++ptr; // skip ' ' after key=value
+        mapping[key] = value;
+    }
+    return mapping;
+}
+
+/** Read full contents of a file and return them in a std::string.
+ * Returns a pair <status, string>.
+ * If an error occured, status will be false, otherwise status will be true and the data will be returned in string.
+ *
+ * @param maxsize Puts a maximum size limit on the file that is read. If the file is larger than this, truncated data
+ *         (with len > maxsize) will be returned.
+ */
+static std::pair<bool,std::string> ReadBinaryFile(const std::string &filename, size_t maxsize=std::numeric_limits<size_t>::max())
+{
+    FILE *f = fopen(filename.c_str(), "rb");
+    if (f == NULL)
+        return std::make_pair(false,"");
+    std::string retval;
+    char buffer[128];
+    size_t n;
+    while ((n=fread(buffer, 1, sizeof(buffer), f)) > 0) {
+        retval.append(buffer, buffer+n);
+        if (retval.size() > maxsize)
+            break;
```

Comment:
This nondeterministically includes all of the last read that exceeded ``maxsize``, rather than truncating to exactly ``maxsize``. Is that intended?